Implementation of Circuit

coxley opened the pull request:
#203 

This is a first iteration. It's pretty weak, I'll admit. Even with a device or interface id, I don't think you can query the circuits like we would want. This is mostly because I lack familiarity with Django, DRF, and the existing  work here.

Ideally I think we would want to be able to do:

```http
GET /api/sites/:id/devices/:id/circuits
GET /api/sites/:id/interfaces/:id/circuits
```

This does work, though, for relating and browsing in the API. I mostly want to get discussion sparked here and get help with the rest.

http://imgur.com/a/Dsftn

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1470,93 @@ def to_dict(self):
+                self.a_endpoint.device.hostname,
+                self.a_endpoint.name,
+            )
+            if getattr(self.z_endpoint.name):
```
There's a `TypeError` here if `self.display_name=None` because the field is currently named `display_name` (not `name`).

coxley replied:
Well, `z_endpoint` is a fk to an interface which is called `name` :D

coxley commented on nsot/models.py:
```diff
@@ -1470,6 +1470,93 @@ def to_dict(self):
+        help_text='Unique ID of the Site this Circuit is under'
+    )
+
+    display_name = models.CharField(
```
Sounds good, will fix

coxley commented:
I have:

* Fixed `set_name` to actually work on save from DRF properly
* Made unique_together better -- still need to probably add a validate_together method to check `a_endpoint` against `z_endpoint` values and vice versa. Not sure if this would make it inefficient though since that'd be extra query at save (foreign keys)

* Changed `display_name` to `name`
* Added detail routes of...


```http

GET /api/site/:id/interfaces/:id/circuit/

GET /api/site/:id/circuits/:id/devices/
GET /api/site/:id/circuits/:id/interfaces/
GET /api/site/:id/circuits/:id/addresses/

```

jathanism commented on nsot/api/views.py:
```diff
@@ -613,6 +613,85 @@ def networks(self, request, pk=None, site_pk=None, *args, **kwargs):
+
+        return self.list(request, queryset=interfaces, *args, **kwargs)
+
+    @detail_route(methods=['get'])
```
For `.addresses()` and `.devices()` detail routes the underlying queries should be defined as model methods/properties on the `Circuit` model, and then called from the view methods. We should try to prevent anything other than basic presentation logic being in the views.

coxley replied:
Fair enough!

coxley replied:
Fixed

jathanism commented on nsot/models.py:
```diff
@@ -28,7 +28,7 @@
 # objects are explicitly NOT USER CONFIGURABLE.
 RESOURCE_BY_IDX = (
-    'Site', 'Network', 'Attribute', 'Device', 'Interface'
+    'Site', 'Network', 'Attribute', 'Device', 'Interface', 'Circuit'
```
Wondering as we grow the base set of resources if there might not be a better way to do this? It's obviously a not a huge deal, but now we've already got `RESOURCE_BY_IDX`, `VALID_ATTRIBUTE_RESOURCES` (which is actually redundant w/ the class names of `Resource` subclasses), and in #231 we've introduced `HOOK_MODELS` that is pretty much also a direct dupe of `RESOURCE_BY_IDX`.

coxley replied:
Would look cleaner to come from a separate diff, but yeah agreed. It's not a huge thing would would safe a couple bits of redundancy.

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1470,105 @@ def to_dict(self):
+        # another z_endpoint, but solves the others
+        unique_together = (
+            ('a_endpoint', 'z_endpoint'),
+            ('a_endpoint',),
```
What's your motivation for this constraint? It would be better represented as a `OneToOneField` relationship for each endpoint which is effectively a `ForeignKey` with a unique constraint but which only returns a single object on either side vs. one-to-many.

coxley replied:
Ohhh, `OneToOneField` is probably what I'm looking for. I'll play with that!

coxley replied:
Fixed, except still needs a validator to make sure existing z_endpoint doesn't get filled in with a_endpoint. Currently can do blah:ae0 <-> blah:ae1 and blah:ae1 <-> blah:ae0

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1470,105 @@ def to_dict(self):
+            z_name = 'None'
+
+        # '--' as delimeter because an intf name may contain '-'
+        self.name = '{}--{}'.format(a_name, z_name)
```
Can we make the delimiter an underscore ;-)?

coxley replied:
Fo' Sho'

coxley replied:
Fixed

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1470,105 @@ def to_dict(self):
+
+    def save(self, *args, **kwargs):
+        self.full_clean()
+        self.clean_fields()
```
Calling `.clean_fields()` is redundant. It's called by `.full_clean()`by default.

coxley replied:
Gotcha

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1470,105 @@ def to_dict(self):
+    # a_endpoint interface
+    a_endpoint = models.ForeignKey(
+        Interface, db_index=True, related_name='circuits_a',
+        null=False, verbose_name='A-side Interface',
```
Why can `a_endpoint` be null but `z_endpoint` cannot?

coxley replied:
Well, it's the other way around. `a_endpoint` should be required but `z_endpoint` MAY be a third-party if the device is on the edge and it's a peering/transit/PNI. Having a `z_endpoint` in that case would require you adding a device and interface for the third party which is kinda unrealistic/unneeded IMO.

coxley commented:
I've added validation for `a_endpoint` and `z_endpoint` to do the cross-over checks.

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1494,131 @@ def to_dict(self):
+
+    def clean_fields(self, exclude=None):
+        self.site_id = self.clean_site(self.site_id)
+        self.a_endpoint = self.clean_a_endpoint(self.a_endpoint_id)
```
`self.name` should be set here so that when a `.full_clean() is performed, the name gets set.

Otherwise this happens (name isn't set):

```python
In [18]: c1 = Circuit.objects.create(a_endpoint=if1, z_endpoint=if2)

In [19]: c1
Out[19]: <Circuit: >
```

Which means the other places where `.set_name()` is called in a `CircuitSerializer.create()` and `CircuitSerializer.update()` can be removed and just rely on validation to cascade correctly.

jathanism commented on nsot/models.py:
```diff
@@ -1470,6 +1494,131 @@ def to_dict(self):
+
+        return value
+
+    def clean_a_endpoint(self, value):
```
This raises the error, and doesn't allow the circuit to be created. But when you lookup the circuit endpoints it still returns a circuit object. Something is screwy.
```
In [12]: c2 = Circuit.objects.create(a_endpoint=if2, z_endpoint=if1)
...
ValidationError: {u'a_endpoint': u'Interface already used as a z_endpoint'}

In [18]: if1.circuit_z
Out[18]: <Circuit: >

In [19]: if1.circuit_a
Out[19]: <Circuit: foo-bar1:eth0_foo-bar2:eth0>

In [20]: if2.circuit_a
Out[20]: <Circuit: >

In [21]: if2.circuit_z
Out[21]: <Circuit: foo-bar1:eth0_foo-bar2:eth0>

In [36]: c2.z_endpoint.id
Out[36]: 1

In [37]: Circuit.objects.all()
Out[37]: [<Circuit: foo-bar1:eth0_foo-bar2:eth0>]

In [38]: Circuit.objects.get(id=1)
---------------------------------------------------------------------------
DoesNotExist                              Traceback (most recent call last)
<ipython-input-38-063bfb293c29> in <module>()
----> 1 Circuit.objects.get(id=1)

/Users/jathan/sandbox/virtualenvs/nsot/lib/python2.7/site-packages/django/db/models/manager.pyc in manager_method(self, *args, **kwargs)
    125         def create_method(name, method):
    126             def manager_method(self, *args, **kwargs):
--> 127                 return getattr(self.get_queryset(), name)(*args, **kwargs)
    128             manager_method.__name__ = method.__name__
    129             manager_method.__doc__ = method.__doc__

/Users/jathan/sandbox/virtualenvs/nsot/lib/python2.7/site-packages/django/db/models/query.pyc in get(self, *args, **kwargs)
    332             raise self.model.DoesNotExist(
    333                 "%s matching query does not exist." %
--> 334                 self.model._meta.object_name
    335             )
    336         raise self.model.MultipleObjectsReturned(

DoesNotExist: Circuit matching query does not exist.
```

jathanism commented on nsot/api/views.py:
```diff
@@ -613,6 +621,71 @@ def networks(self, request, pk=None, site_pk=None, *args, **kwargs):
+            cir = serializers.CircuitSerializer(interface.circuit)
+            return self.success(cir.data)
+        except models.Circuit.DoesNotExist:
+            msg = 'No Circuit found at Interface (site_id, id) = ({}, {})'
```
BTW this just a comment for now.

coxley commented on nsot/models.py:
```diff
@@ -1470,6 +1494,131 @@ def to_dict(self):
+    # a_endpoint interface
+    a_endpoint = models.OneToOneField(
+        Interface,
+        on_delete=models.CASCADE,
```
Probably want to disallow deletion, so `PROTECT` ?

coxley commented on nsot/models.py:
```diff
@@ -1267,6 +1278,19 @@ def networks(self):
+    @property
+    def circuit(self):
+        """Return the Circuit I am associated with"""
+        try:
```
Yeah, I'm not sure how we can make it a single lookup

coxley commented on nsot/models.py:
```diff
@@ -1470,6 +1494,131 @@ def to_dict(self):
+
+    @property
+    def addresses(self):
+        interfaces = self.interfaces
```
Yeah, maybe the Django ORM can support that lookup. Could be done with SQL pretty easily.

jathanism commented:
I've taken over this feature and will be replacing it with a new PR.
